### DISCUSSION TITLE
When leaving Android game via back button (triangle), you can't always resume - it re-exits

### DISCUSSION THREAD

**Discussion by derekschultz0** [MonoGame/MonoGame#8440] [Android]
Hi All,

I am using MonoGame 3.8.1 (latest from NuGet) and have noticed with my Android app, that if I use the Android Back button to exit (go back to the home page, but not actually quit/kill the app), that if I return, it often will reclose right after opening.  Maybe my fatal flaw is just allowing this, which works the same as the Home button (which works just fine).  I certainly can do a kill on the app and it works as expected there.

I tried some other workarounds, but they didn't seem to do the trick for me.

When debugging, the thing I noticed was when I try to resume, I get the OnActivate event, but when it goes to the Update, it thinks the Back button is still pressed, thus re-exits.  When I look at the code, it seems that the GamePad.Back flag is set to true when pressed, but only gets set back to false when the OnKeyUp event for the Back button is handled.  Apparently that doesn't always happen.  So you get in a loop with the app always thinking the back button is pressed.

I have a workaround - a kludgy workaround, but it seems to work - where I can call OnKeyUp to reset the Back button to false.  

If there is a better way, please let me know.  If this really is an issue, maybe some future release can better handle the back button, or expose it in a way for a direct clearing.

Here is the workaround.  Thanks.

```
private MonoGameAndroidGameView _monoGameAndroidGameView = null;
private KeyEvent _dummyKeyEvent = null;

protected override void Update(GameTime gameTime)
{
	if (GamePad.GetState(PlayerIndex.One).Buttons.Back == ButtonState.Pressed)
	{
		try
		{
			if (_monoGameAndroidGameView == null)
			{
				// NOTE: GamePad.Back is STATIC, so we can get away with separate instance referencing it
				_monoGameAndroidGameView = new MonoGameAndroidGameView(Game.Activity, (AndroidGameWindow)Window, this);
			}

			if (_dummyKeyEvent == null)
			{
				// create dummy keyevent so we can call OnKeyUp
				_dummyKeyEvent = new KeyEvent(0, "", 0, KeyEventFlags.Canceled);
			}

			if (_monoGameAndroidGameView != null)
			{
				try
				{
					// Reset our BACK button flag so we can resume on activation, without re-BACKing out...
					// NOTE: SIDE EFFECT of call is setting GamePad.Back = false - so it is NOT stuck on true, which causes exit loop on reactivate
					_monoGameAndroidGameView.OnKeyUp(Keycode.Back, _dummyKeyEvent); 
				}
				catch
				{
					// ignore error (if any) - side effect from above call will still take effect
				}

				Game.Activity.MoveTaskToBack(true);
				base.Update(gameTime);
			}
			else
			{
				throw new Exception("Unable to clear BACK button"); // below catch will do hard exit to prevent loop
			}
		}
		catch
		{
			// unable to close the way we want - just kill the process
			Game.Activity.MoveTaskToBack(true);
			Process.KillProcess(Process.MyPid());
		}
	}
	else
	{
		base.Update(gameTime);
	}
}
```

**Comment by stromkos** [comment]
You have a memory reference that is left hanging somewhere.


In order to catch the back (triangle) button you must add a listener for the native back event to the `Activity.cs` file. 

The app never has a chance to process the background call (60 fps vs real-time event).


If your game is removed to make space(memory) for another task, all state will be lost.

**Reply by derekschultz0** [reply]
I appreciate you taking the time to reply.  

It is certainly possible I have a memory issue somewhere.  I did make an attempt to add a OnBackPressed() event to try to intercept the call in activity1.cs.  I never actually hit that breakpoint, so I'm probably doing something wrong there.  One thing I did do, is create a brand new project from the monogame android template.  I configured it to .net 8, target os 34.0, supported os 23.0.  

When I run the project I get the typical CornFlowerBlue screen.  If I click the back button (triangle), it does close.  The first time I tried restoring it worked, but the 2nd time I tried it did not - i.e. it just re-closed quickly.  Again, I tried the OnBackPressed() event, but never hit the breakpoint.

Do you have a small, maybe just a template example for MonoGame 3.8.1, .net 8, that you can share?  Again thanks for any suggestions.

**Reply by stromkos** [reply]
Don't try to handle the "back" button at all.  


Mobile users don't click on an X to close a program. They hit the task `|||`  button to open a new or different task. The close comes with the swipe or close all (or stop task from Apps).

The OS and MonoGame will handle the task events: background, swap and resume for you.

If there is any state to be saved provide a checkpoint, button or timed save and restore on load. 

You could hook the suspend and resume events to save and restore, if more granular save timing is needed.

---

> It is certainly possible I have a memory issue somewhere.

Please do not look for one.  I was talking about the OS.

---

My games stay loaded and resume-able until the OS needs the resources at that point the unsaved progress is lost, and the game is reloaded on restore.

**Reply by derekschultz0** [reply]
Ok, but even with the standard monogame android template without any customization, if I click the back button, just using built-in behavior, if I then resume, it may work a time or two, then I can't resume ever - it just flashes the screen and exits again.  The only way then to load the app and keep it open is to do a force unload, then do it again.  I've seen this with 2 different moto phones and also the pixel 5 emulator.

**Reply by stromkos** [reply]
What Android API version?  "pixel 5" is at API level 11.

18 is the minimum Google requires for security.

**Reply by derekschultz0** [reply]
My system shows Pixel 5 API 30 (Android 11.0 - API 30)

**Reply by dellis1972** [reply]
I would give this a test myself but I don't have a modern device with a hardware back button . 
Generally though on Android its down to the operating system to decide when to kill the app, like on iOS. 
Can you get an `adb logcat` of the output when you get the issue in the template? There might be some diagnostic output we can use to figure out the issue.

**Reply by derekschultz0** [reply]
Hi dellis1972, I will look into adb logcat.  My phone actually doesn't have a hardware buttons, but uses the software buttons that appear when you swipe.  On the Pixel5 emulation, I can use the swipe approach or the command bar buttons. Ultimately, when I tried to find the issue myself, I saw that the only thing that fixes the  GamePad.Back flag, is when OnKeyUp for the back button is processed, but apparently this doesn't always get called, leaving MonoGame to think the back button is still pressed, long after the fact.  But I will try to look more at this, but can't today due to other commitments.  Thank you for your interest in trying to supersleuth this situation.

**Reply by derekschultz0** [reply]
The approach I've seen some people take, is to forcibly kill the process, which does work - just not sure I want to do that.  Again, thanks.
